src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIMarsConnection.cs

```diff
@@ -0,0 +1,314 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+namespace System.Data.SqlClient.SNI
+{
+    /// <summary>
+    /// SNI MARS connection. Multiple MARS streams will be overlayed on this connection.
+    /// </summary>
+    internal class SNIMarsConnection
+    {
+        private SNIHandle _lowerHandle;
+        private ushort _nextSessionId = 0;
+        private Guid _connectionId = Guid.NewGuid();
+        private Dictionary<int, SNIMarsHandle> _sessions = new Dictionary<int, SNIMarsHandle>();
+        private byte[] _headerBytes = new byte[SNISMUXHeader.HEADER_LENGTH];
+        private int _currentHeaderByteCount = 0;
+        private int _dataBytesLeft = 0;
+
+        private SNISMUXHeader _currentHeader;
+        private SNIPacket _currentPacket;
+
+        /// <summary>
+        /// Connection ID
+        /// </summary>
+        public Guid ConnectionId
+        {
+            get
+            {
+                return _connectionId;
+            }
+        }
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="lowerHandle">Lower handle</param>
+        public SNIMarsConnection(SNIHandle lowerHandle)
+        {
+            _lowerHandle = lowerHandle;
+            _lowerHandle.SetAsyncCallbacks(HandleReceiveComplete, HandleSendComplete);
+        }
+
+        public SNIMarsHandle CreateSession(Object callbackObject, bool async)
+        {
+            lock (this)
+            {
+                ushort sessionId = _nextSessionId++;
+                SNIMarsHandle handle = new SNIMarsHandle(this, sessionId, callbackObject, async);
+                _sessions.Add(sessionId, handle);
+                return handle;
+            }
+        }
+
+        /// <summary>
+        /// Start receiving
+        /// </summary>
+        /// <returns></returns>
+        public uint StartReceive()
+        {
+            SNIPacket packet = null;
+
+            if (ReceiveAsync(ref packet) == TdsEnums.SNI_SUCCESS_IO_PENDING)
+            {
+                return TdsEnums.SNI_SUCCESS_IO_PENDING;
+            }
+
+            SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, 0, "Could not start MARS async receive");
```

Comment:
I'm used to file I/O where an async operation can actually return synchronously, in which case it won't return "pending".  That's not the case here, right?  ReceiveAsync will return PENDING even if it actually completed synchronously (or maybe synchronous completion isn't possible)?